tidy3d/components/data/monitor_data.py

```diff
@@ -374,6 +441,114 @@ def dot(
+
+        See also
+        --------
+        :member:`dot`
+        """
+
+        tan_dims = self._tangential_dims
+
+        # pylint: disable=protected-access
+        assert all(
+            a == b for a, b in zip(tan_dims, field_data._tangential_dims)
+        ), "tangential dimensions must match"
+
+        # Tangential fields for current
+        fields_self = self._centered_tangential_fields
+        if conjugate:
+            fields_self = {component: field.conj() for component, field in fields_self.items()}
+
+        # Tangential fields for other data
+        # pylint:disable=protected-access
+        fields_other = field_data._interpolated_tangential_fields(self._plane_grid_centers)
+
+        # Tangential field component names
+        dim1, dim2 = tan_dims
+        e_1 = "E" + dim1
+        e_2 = "E" + dim2
+        h_1 = "H" + dim1
+        h_2 = "H" + dim2
+
+        # Prepare array with proper dimensions for the dot product data
+        arrays = (fields_self[e_1], fields_other[e_1])
+        coords = (arrays[0].coords, arrays[1].coords)
+        # Common frequencies to both data arrays
+        f = np.array(sorted(set(coords[0]["f"].values).intersection(coords[1]["f"].values)))
+        mode_index_0 = coords[0]["mode_index"].values
+        mode_index_1 = coords[1]["mode_index"].values
+        dtype = np.promote_types(arrays[0].dtype, arrays[1].dtype)
+        dot = np.empty((f.size, mode_index_0.size, mode_index_1.size), dtype=dtype)
+
+        # Calculate overlap for each common frequency and each mode pair
+        for i, freq in enumerate(f):
+            for mi0 in mode_index_0:
+                e_self_1 = fields_self[e_1].sel(f=freq, mode_index=mi0, drop=True)
+                e_self_2 = fields_self[e_2].sel(f=freq, mode_index=mi0, drop=True)
+                h_self_1 = fields_self[h_1].sel(f=freq, mode_index=mi0, drop=True)
+                h_self_2 = fields_self[h_2].sel(f=freq, mode_index=mi0, drop=True)
+
+                for mi1 in mode_index_1:
+                    e_other_1 = fields_other[e_1].sel(f=freq, mode_index=mi1, drop=True)
+                    e_other_2 = fields_other[e_2].sel(f=freq, mode_index=mi1, drop=True)
+                    h_other_1 = fields_other[h_1].sel(f=freq, mode_index=mi1, drop=True)
+                    h_other_2 = fields_other[h_2].sel(f=freq, mode_index=mi1, drop=True)
+
+                    # Cross products of fields
+                    e_self_x_h_other = e_self_1 * h_other_2 - e_self_2 * h_other_1
+                    h_self_x_e_other = h_self_1 * e_other_2 - h_self_2 * e_other_1
+
+                    # Integrate over plane
+                    d_area = self._diff_area
+                    integrand = (e_self_x_h_other - h_self_x_e_other) * d_area
+                    dot[i, mi0, mi1] = 0.25 * integrand.sum(dim=d_area.dims)
+
+        coords = {"f": f, "mode_index_0": mode_index_0, "mode_index_1": mode_index_1}
```

Comment:
I guess this is fine, but there's still quite some overlapping code between this method and `dot`. Did you have any particular reason against going the route of using (custom) `isel` type methods and direct calls to `dot`?

Reply:
There was the issue with setting `colocate = True` for `ModeSolverMonitor`, which doesn't work and, tbh, creating a separate monitor and doing custom selections to get `dot` to perform the operation between the desired modes seemed like increasing complexity to me, compared to just doing the operation in place. I would actually consider the other way around, having `dot` as a special case of `outer_dot`, but then we also lose performance doing unnecessary interpolations. In the end (besides the monitor issue) it just felt like the code duplication here is minimal compared to complexity cost, but I can change that if you don't want the duplication here. If that's the case, can I just make a `ModeMonitor` from a `ModeSolverMonitor`?

Reply:
Ok, let's leave it like this for now, maybe I'll rethink it later when making the changes to colocate by default.